### DISCUSSION TITLE
Lock CRON jobs when using multiple instances

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#13381] [Feature Requests]
At the moment CRON jobs are possible with extensions.
However if you have multiple Directus instances running this will result in multiple CRON jobs running at the same time.
For example:
You have a CRON job, it creates and sends invoices based on worked hours in the last month.
But you have 2 instances so the CRON job will get fired twice.

A possible solution:
Whenever a Directus instance connects to Redis, make a new record, for example CRONlocked:true
Then whenever a second instance connects to Redis, it would not enable CRON jobs for that instance.
If the Directus instance goes offline that setted the key, it will clear the CRONlocked

We could also just create a lock file or something on the disc when redis is not enabled.

**Comment by azrikahar** [comment]
cc @nickrum & @br41nslug based on our recent discussions which does (albeit not entirely) involve this topic!

**Comment by nickrum** [comment]
This is especially relevant with Data Flows being almost ready to merge.

The solution you are describing is a very basic "leader election" algorithm. The tricky part with this approach would be what to do when the node that is responsible for executing CRON jobs crashes? In that case the lock has to somehow be automatically unlocked.

The approach we are currently pursuing is a little bit more involved than that. It will most likely also be using redis to "synchronize" instances, but it will use a logical clock to keep track of CRON job executions alongside a locking mechanism while the CRON job is executing.

**Comment by u12206050** [comment]
I implemented a queue and jobs scheduler which solves this for me as well as gives me over view in the admin the status of the jobs.

<img width="1291" alt="Screenshot 2022-05-19 at 19 30 41" src="https://user-images.githubusercontent.com/6641242/169361798-cae6d1f1-94e7-4496-b7a2-346a50aba903.png">

The way it works is every time a job (any async action incl. cron jobs) needs to execute it gets upserted to the database with status pending. Every minute a cron job runs, be it on 1 or multiple containers, they query the `jobs` table for a random pending job.

So in one db transaction I select and update a random job to status `processing` together with the container id, so I can track how many jobs a single container was responsible for.

The job is then executed and once completed (or failed) the job record is updated to reflect the state.

Every job has a `payload` as well as an `action`. It is the `action` that defines what function will be called and passed the arguments.

Benefits of this as mentioned before, is:
 - total insight into running, pending and completed jobs.
 - Able to timeout and retry jobs if a container were to fail.
 - Delay certain jobs to only run after X amount of time or set date.

No need to sync across instances, since the db transaction ensures only one instance executes a particular job.
The random selection is just too minimise contention when picking a job to do.
Meaning you can execute jobs in parallel using multiple instances.

**Reply by u12206050** [reply]
Oh yes, and if a particular job gets queued multiple times, it still only gets queued once, for a unique payload set, since it is assumed the outcome would be the same nonetheless. ie. Recalculate the total number of items in list X.

However this does mean I keep track of a queue_at date, that I can use to debounce before executing.

**Comment by gdpdils** [comment]
This is much needed. As it is, the implementation is almost doomed to be useless in a real-life scenario and I don't think users should develop ad-hoc solutions for something that should be integrated into the framework.

**Reply by benhaynes** [reply]
Yes, this is something we are working to include in the core service. @nickrum any updates on this one?

**Comment by rijkvanzanten** [comment]
This is being worked on and can be tracked through https://github.com/directus/directus/issues/15052
